pallets/xvm/src/lib.rs

```diff
@@ -227,11 +250,26 @@ impl<T: Config> Pallet<T> {
             T::Lookup::lookup(decoded).map_err(|_| error)
         }?;
 
+        // Adjust `value` if needed, to make sure `source` balance won't be below ED after calling
+        // a payable contract. This is needed as `pallet-contracts` always respects `source` ED.
+        // Without the adjustment, the first call from any `source` to a payable contract will
+        // always fail.
+        //
+        // Only the first call to a payable contract results less `value` by ED amount, the following
+        // ones will not be affected.
+        let source_balance = T::Currency::free_balance(&source);
+        let existential_deposit = T::Currency::minimum_balance();
+        let killing_source = source_balance.saturating_sub(value) < existential_deposit;
+        let adjusted_value = if killing_source {
+            value.saturating_sub(existential_deposit)
+        } else {
+            value
+        };
```

Comment:
Why not just let it fail/revert?
Shouldn't `pallet-contract` (and `pallet-balances`) already ensure this?

Reply:
I think we should fail the call if this happens, otherwise for users it will be an unexpected result.

If I understand correctly, Let's say ED=10, EVM Contract balance is 15 and it tries to send 10 to Wasm contract (nft mint), but 0  (10 - 10) will be sent? That does not seems appropriate

Also I don't quite understand this "_Only the first call to a payable contract results less `value` by ED amount, the following ones will not be affected._", what does that mean?

Reply:
This is addressed in the integrate test. https://github.com/AstarNetwork/Astar/blob/22dc6e2d20ce24e6a88b75c73f689568bf26443c/tests/integration/src/xvm.rs#L224

If we let it fail/revert, payable calls from EVM to WASM always fail, until the EVM contract gets some balance above ED before the call. The problem is that EVM contracts start with zero balance, but for payable calls, `pallet-contracts` will assume transferring the full amount `value` will kill this EVM contract, and it fails with `TransferFailed` error. Tho `pallet-contracts` is wrong, an EVM contract can stay at zero balance.

Reply:
For Ash's comment on the unexpected result, IMO ED is typically set reasonably low that it doesn't matter.

In your example, if you mean the EVM contract has `15` after its payable function called, `5` will be sent instead of `0`, otherwise if it's before the payable function called, the full amount is transferred.

*Only the first call...* means that after a payable call from EVM to WASM is successfully made, this contract will have enough balance for ED because of this adjustment. So for following calls, the full amount will always be transferred.

Reply:
I understand now, thanks for explaining :)
In nutshell, we just debit ED from value if evm contract does not have it, that's one time cost after that evm contract has ED so pallet-contracts will not complain.
That means first transfer should be greater than ED, as you said ED is quite low so not a issue, but we should document it somewhere for ourselves and builders

Reply:
Correct. It's a one-time cost. Since it's just ED value loss, I believe users won't mind.

In the last sync with Ivan and Pierre, we agreed I'll write documents about XVM architecture after most XVM features are done. I'll add this to it.

Reply:
@PierreOssun Yes, it will fail because of below ED, as EVM also use `T::Currency::transfer` which will check dest ED. Btw Frontier maps all transfer error into `OutOfFund`.

https://github.com/paritytech/frontier/blob/22aaafe089218f6cee625898fff7b953cc793228/frame/evm/src/runner/stack.rs#L909

Reply:
I've gone over the comments a few times, and must admit I'm a bit confused. 🙂 

The way I see this:
* `Alith` calls an EVM contract
* this call, internally, calls a precompile with a `value` transfer
* `source` will be the contract, not `Alith`, so any _payable value_ is deducted from the contract

In case EVM contract doesn't have funds, then it cannot make the `payable` call.
EVM contract address is same as any other, so it has to have at least the ED if it has _any_ funds.

The EVM call `Alith` makes doesn't even have to be payable. But it can internally call a `payable` function.

So prerequisite for EVM contract to make a `payable` call should be to have some funds.

Maybe this is a dumb comment, but it's how I see things 😅

Reply:
> So prerequisite for EVM contract to make a `payable` call should be to have some funds.

Yes, this is true. The source of the fund could be: 1. This *making a payable call* piece of code is embedded in a payable call, which could be called and funds transferred to the contract. 2.  The contract has other payable calls, and received funds via them.

The `value` adjustment here is for the first case.

> Maybe this is a dumb comment, but it's how I see things 😅

I think it is worth to further elaborate. Let's use the integration test mentioned above as an example.

1. We deploy an EVM contract as follows. It has a payable function calling XVM precompile, which is also a payable function. Let's call the contract `EP`. https://github.com/AstarNetwork/Astar/blob/dda5e361b191fd66eb5c074d859f5c8d307ddbb5/tests/integration/src/xvm.rs#L88-L92
2. We deployed a WASM contract, it has a payable `deposit` function as follows. Let's call the contract `WP`.
https://github.com/AstarNetwork/Astar/blob/dda5e361b191fd66eb5c074d859f5c8d307ddbb5/tests/integration/src/xvm.rs#L66-L67
3. After the deployment, `EP` has a balance of 0.
4. When `Alilth` make the call to `EP` with 1 ASTR as `value`:
    - 1 ASTR is transferred and `EP` has a balance of 1 ASTR.
    - In this call, `EP` calls precompile, which then calls `WP`'s payable call with 1 ASTR `value` via XVM.
    - With `WP`'s payable call is called, `pallet-contracts` will try to transfer 1 ASTR from `EP` to `WP`. And this transfer causes the issue:
        - Without the value adjustment, `EP` will have a balance of 0 after the transfer. As `pallet-contracts` assumes this will kill `EP` (which ofc is not true), it fails the transfer with `TransferFailed` error. End of story. The cross-VM payable call failed. But this shouldn't be the case, because apparently `pallet-contracts` is wrong.
        - With the value adjustment, `EP` will have a balance of ED after the transfer. `pallet-contracts` is happy to see `EP` will be alive, and the transfer is successful. `WP` gets `1 ASTR - ED`. The cross-VM payable call is finished.

Reply:
Thanks a lot for the detailed explanation!
I checked all the parts, including the integration test, before posting the question. 🙂 
What I missed was that this is focused on a very particular case. It was in the example, I just didn't see it, my bad.

My opinion is still we should let it fail.
Let me share why:
* I guess we can assume users always send the _correct_ amount of `value` via the `payable` call
  * if that assumption holds, then the destination for the funds won't receive what it should have
  * this might result in failure (insufficient funds for some operation, e.g. NFT mint) or unexpected behavior (e.g. image you're bidding on something and you loose by ED 😂 )
* `value` change is not common in any of the VMs/SC engines AFAIK
* this puts the "burden" on the user

Instead, I'd recommend to just document the requirement and make it clear to the devs about this particular situation. If they have such a call in their code, they should make sure that EVM smart contract address has ED.
This puts the "burden" on devs.

Just my suggestion.

Reply:
I understand your concern. From a pure engineering point of view, I agree we should just let it fail as it is. This solution is even simpler to implement. Tho practically, I'm more concerned with the extra friction for adoption.

From my experience, it's quite common for one payable EVM function to call another payable, for instance in DEX. EVM doesn't have the concept of ED, and we'll need to educate Solidity builders about ED and substrate `pallet-contracts`. Then they have to pay attention to any XVM payable usage, or it just won't work. Like our (rather complicated) underlying account system, it will be another burden for them to understand. IMO this kind of friction should be avoided by design.

To be fair, some edge cases you mentioned would be a concern, but compared to the case "dev forgot to load up another contract using XVM payable after new factory deploy or upgrade, then some features won't work", I don't know which is worse. And it's like `Unlimited` parameter in XCM: do we prefer to fail all the tx if the weight limit is not, until we do a runtime upgrade to fix, or by using `Unlimited` the tx is always successful but users might receive less value?

About the reason why other VMs don't have similar stuff like `value` changes, I believe it's more because their VM doesn't have to bridge two systems with different designs. They don't have to deal with account unification either😂. (Very lucky! cc @ashutoshvarma ).

Reply:
I can't say which solution is better, as apparently, I'm biased as a human. The fairest way to decide is probably trying it in real life.

To push this PR and XVM work forward, a future-proof way might be: remove the `value` adjustment and let devs feel the burden. If no Solidity builder is using XVM payable calls, then nobody feels the burden, which is very good. If they find the burden is okay, then nobody feels the pain, good. Or if they find it's quite a friction for adoption and we find it's hard to let them understand the underlying gap, it's never too late to update XVM with `value` adjustment.

Reply:
Thanks for the input. I'll update to remove the `value` adjustment for now. And let's mark this as _open to discussion with builders_. We are building this for them to use anyway. Let's see what the feedback is after they learn about it.

If they don't like it, let's then discuss the solution. There could be other possible options, for instance, we can load up `pallet-xvm`'s pallet account with funds from Astar, and transfer ED amount to EVM contracts automatically if needed.

Reply:
I'd same concerns regarding this approach, but thinking back again it's not a big deal since it will happen only for the _very first payable call via tha EVM contract_, and most of the time it will be the developers of contract themselves calling it for the first time (testing, staging env).
So for users it will not be a very big burden to say (except for the very first one 😆)